x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregatesWithNull.java

```diff
@@ -46,48 +45,44 @@ public ReplaceAggregatesWithNull() {
+        List<NamedExpression> remainingAggregates = new ArrayList<>(aggregates.size());
         boolean changed = false;
-        List<Alias> transientEval = new ArrayList<>();
+        List<Alias> replacingEvalVariables = new ArrayList<>();
 
-        for (NamedExpression agg : aggs) {
+        for (NamedExpression agg : aggregates) {
             Expression e = Alias.unwrap(agg);
             Object value = null;
-            boolean isNewAgg = false;
-            if (e instanceof AggregateFunction af
-                && af.field().foldable()
-                && (DataType.isNull(af.field().dataType())
-                    || af.field() == null
-                    || af.field() instanceof Literal lit && lit.value() == null)) {
-                isNewAgg = true;
+            boolean isNullAgg = false;
+            if (e instanceof AggregateFunction af && Expressions.isNull(af.field())) {
+                isNullAgg = true;
                 if (af instanceof CountDistinct || af instanceof Count) {
                     value = 0L;
                 }
-            } else if (e instanceof Literal literal && literal.value() == null) {
-                isNewAgg = true;
+            } else if (Expressions.isNull(e)) {// FoldNull can replace an entire aggregate function with a null Literal
+                isNullAgg = true;
             }
 
-            if (isNewAgg) {
+            if (isNullAgg) {
                 /*
                  * Add an eval for every null (even if they are all "null"s, they can have different data types,
                  * depending on the Aggregate function return type).
                  * Also, copy the original alias id so that other nodes using it down stream (e.g. eval referring to the original agg)
                  * don't have to be updated. PruneColumns makes use of the Attribute ids to decide if unused references can be removed.
                  */
                 var aliased = new Alias(agg.source(), agg.name(), Literal.of(agg, value), agg.toAttribute().id(), true);
-                transientEval.add(aliased);
+                replacingEvalVariables.add(aliased);
                 changed = true;
             } else {
-                newAggs.add(agg);
+                remainingAggregates.add(agg);
             }
         }
 
         LogicalPlan plan = aggregate;
-        if (changed) {
+        if (changed) {// if there were null aggregates found
             var source = aggregate.source();
-            if (newAggs.isEmpty() == false) {
-                plan = new Aggregate(source, aggregate.child(), aggregate.aggregateType(), aggregate.groupings(), newAggs);
+            if (remainingAggregates.isEmpty() == false) {// build the new Aggregate with the rest (non-null) aggregates
+                plan = new Aggregate(source, aggregate.child(), aggregate.aggregateType(), aggregate.groupings(), remainingAggregates);
             } else {
                 // All aggs actually have been optimized away
```

Comment:
I am not following and I need more details. `remainingAggregates` being empty means exactly that - the `BY` part is empty, it was folded no `null` by this rule. If `BY` is reduced to `null` this means only one row in the results.

Reply:
Actually, I found a reproducer on your branch:
```
ROW field = null::integer, otherfield = [1,2] | STATS min(field) by otherfield | DROP otherfield

  min(field)   
---------------
null 
```
This should instead be
```
  min(field)   
---------------
null 
null
```
(as it is on main) because there 2 values for `otherfield` and thus 2 groups, resp. 2 output rows from the `STATS`.

It's insufficient to look at the aggregates. We need to check `aggregate.groupings()` instead, as dropping the groupings after the stats lets us optimize them away from the aggregates.

The removal of the groupings from the aggregates happens in `CombineProjections`, like in the example above:
```
[2024-10-17T11:25:35,813][INFO ][o.e.x.e.o.LogicalPlanOptimizer] [runTask-0] Rule logical.CombineProjections applied
Limit[1000[INTEGER]]                                                                                     = Limit[1000[INTEGER]]
\_EsqlProject[[min(field){r}#7]]                                                                         ! \_Aggregate[STANDARD,[otherfield{r}#4],[MIN(field{r}#2,true[BOOLEAN]) AS min(field)]]
  \_Aggregate[STANDARD,[otherfield{r}#4],[MIN(field{r}#2,true[BOOLEAN]) AS min(field), otherfield{r}#4]] !   \_Row[[TOINTEGER(null[NULL]) AS field, [1, 2][INTEGER] AS otherfield]]
    \_Row[[TOINTEGER(null[NULL]) AS field, [1, 2][INTEGER] AS otherfield]]                               ! 
```

Reply:
I don't think `CombineProjections` is to blame, it is doing the right thing. This can be seen with queries that have their projections correctly pruned away.

The "workaround" of calling `SubstitueSurrogates` again in the "Operator Optimization" is to blame here and this shows the vicious cycle of fixing something - breaking something else that [I mentioned before](https://github.com/elastic/elasticsearch/pull/112392/files#r1802734728). Foldability and surrogate expressions represent a faulty principle and when things start to not make sense and the code "struggles" to make the right thing (by using workarounds like calling `SubstituteSurrogates` twice) then it is clear that something more fundamental is not right.

At this point there are two options available:
- create an issue for this and merge the PR for the benefits of the tests it adds. This though introduces a bug in a edgy situation (the one that we know of, there may be others that we don't)
- try to make `SubstituteSurrogates` don't do its thing for aggregations that have a grouping which is not kept as a projection anymore. This is clearly a workaround that tries to ignore what `CombineProjections` did. I am trying now this second idea.